### DISCUSSION TITLE
A strange glitch with trying to programmatically pop a `NavigationLinkStore`

### DISCUSSION THREAD

**Discussion by oliverfoggin** [pointfreeco/swift-composable-architecture#1996] [Beta]
I initially had an issue with `NavigationLinkStore` as I was observing too much information. I feel like I am now observing the right data but I'm unsure.

I have been working to pare that back and think I have done it but now I've encountered another issue.

I've managed to replicate it again in a small project and have a video.

At the "Start" of the app I have a `NavigationLinkStore` to "Go to level 1 >". This creates state for the next level and the `NLS` pushes the next view on.

Then on each subsequent "level" I have two buttons. One to go to the next level, and one to pop back.

If I pop back at the first level it works. But after that it fails. It nils out the state but the pop doesn't happen. The pop happens by the parent picking it up and nilling out the state.

It is a contrived example but it is representative of what is happening in my own app too.

Here is the view and reducer

```
import SwiftUI
import ComposableArchitecture

struct Level: Reducer {
	struct State: Equatable {
		@PresentationState var nextLevel: Level.State?
		var level: Int
		var title: String { "Go to level \(level + 1) >" }
	}

	indirect enum Action: Equatable {
		case onAppear
		case levelTapped
		case pop

		case nextLevel(PresentationAction<Level.Action>)
	}

	var body: some ReducerOf<Self> {
		Reduce<State, Action> { state, action in
			switch action {
			case .onAppear:
				print("On appear level \(state.level)")
				return .none

			case .levelTapped:
				state.nextLevel = .init(level: state.level + 1)
				return .none

			case .pop:
				return .none

			case .nextLevel(.presented(.pop)):
				state.nextLevel = nil
				return .none

			case .nextLevel:
				return .none
			}
		}
		.ifLet(\.$nextLevel, action: /Action.nextLevel) {
			Level()
		}
	}
}

struct LevelView: View {
	let store: StoreOf<Level>

	struct ViewState: Equatable {
		let title: String
		let level: Int

		init(state: Level.State) {
			title = state.title
			level = state.level
		}
	}

	var body: some View {
		WithViewStore(store, observe: ViewState.init) { viewStore in
			VStack(spacing: 10) {
				NavigationLinkStore(
					store.scope(state: \.$nextLevel, action: Level.Action.nextLevel)
				) {
					viewStore.send(.levelTapped)
				} destination: { store in
					LevelView(store: store)
				} label: {
					Text(viewStore.title)
				}

				Button {
					viewStore.send(.pop)
				} label: {
					Text("< Pop back to level \(viewStore.level - 1)")
				}
			}
			.onAppear {
				viewStore.send(.onAppear)
			}
			.navigationTitle("Level \(viewStore.level)")
		}
	}
}
```

Here is a video of what I am seeing.

https://user-images.githubusercontent.com/1813244/227010602-d899b20a-d858-4424-a416-54739b492a96.mov

As you can see I can pop back from "Level 1" to the start. But if I try to pop back from anything beyond the first level it doesn't pop even though the state is nilled out.

**Comment by stephencelis** [comment]
I think you may be butting heads with general SwiftUI bugs. The old `NavigationLink` (which `NavigationLinkStore` wraps) is a bit buggier than the newer APIs (and the newer APIs are buggy, too!). One thing we've noticed is that you must explicitly set `navigationViewStyle(.stack)` on iOS to avoid some issues with linking over multiple layers, but also, linking over multiple layers itself can be buggy, and require you to programmatically present A before B, and programmatically dismiss B before A, with some kind of delay in between.

**Reply by oliverfoggin** [reply]
Thanks for the reply.

I found that adding the `navigationViewStyle` worked a treat 👍🏻 

Plus, I also had a play around and found that if I just used the `@Environment(\.dismiss)` from the view that also worked and even sent the dismiss action into the store. Not ideal but it works.

Thanks again for your help 😄
